src/app.py

```diff
@@ -1263,6 +1269,545 @@ def allowable_edit_states(hmuuid):
+            fcount = fcount + 1
+            temp_file_name = tfile
+        incorrect_file_count = False
+        incorrect_file_count_message = ""
+        if fcount == 0:
+            #raise Exception("File not found for temporary file with id " + temp_id) #how are these getting passed to the front end?
+            incorrect_file_count = True
+            incorrect_file_count_message = f"File not found in temporary directory /{temp_id}"
+        if fcount > 1:
+            #raise Exception("Multiple files found in temporary file path for temp file id " + temp_id)
+            incorrect_file_count = True
+            incorrect_file_count_message = f"Multiple files found in temporary file path /{temp_id}"
+        if incorrect_file_count:
+            return_response = {"status": "fail", "message": incorrect_file_count_message}
+            return Response(json.dumps(return_response, sort_keys=True), 400, mimetype='application/json')
+        if incorrect_file_count is False:
+            tsvfile_name = tsv_directory + temp_file_name
+            records = []
+            headers = []
+            with open(tsvfile_name, newline= '') as tsvfile:
+                reader = csv.DictReader(tsvfile, delimiter='\t')
+                first = True
+                for row in reader:
+                    data_row = {}
+                    for key in row.keys():
+                        if first:
+                            headers.append(key)
+                        data_row[key] = row[key]
+                    records.append(data_row)
+                    if first:
+                        first = False
+            validfile = validate_donors(headers, records)
+            if type(validfile) == list:
+                return_validfile = {}
+                error_num = 0
+                for item in validfile:
+                    return_validfile[str(error_num)] = str(item)
+                    error_num = error_num + 1
+                # return_validfile = ', '.join(validfile)
+                response_body = {"status": "fail", "data": return_validfile}
+                return Response(json.dumps(response_body, sort_keys=True), 400, mimetype='application/json')
+            entity_response = {}
+            row_num = 1
+            if validfile == True:
+                for item in records:
+                    item['lab_donor_id'] = item['lab_id']
+                    del item['lab_id']
+                    item['label'] = item['lab_name']
+                    del item['lab_name']
+                    item['protocol_url'] = item['selection_protocol']
+                    del item['selection_protocol']
+                    if include_group:
+                        item['group_uuid'] = group_uuid
+                    r = requests.post(commons_file_helper.ensureTrailingSlashURL(app.config['ENTITY_WEBSERVICE_URL']) + 'entities/donor', headers=header, json=item)
+                    entity_response[row_num] = r.json()
```

Comment:
Missing exception handling here for the `r` response. We need to check based on the status code, if 200, you can call `r.json()` to handle the result. Otherwise, need to throw error message.

Reply:
this is already kind of "handled". Bill and I were going back and forth on what we want to happen in the event that some of the entries in a tsv file were successfully created by entity api, and some were not. We decided it should still return a 201 as something was created. The way it currently works: through the loop, each time entity api is called, whether successful or unsuccessful, the response from entity api gets recorded and returned from my method. 

if a tsv file has 4 donors and 3 of them are created successfully and 1 of them has an error (for example, lab_name already exists or something), my method should return the response from entity for each row. So there will be 3 201's and 1 400. 

So we don't want entity_response[row_num] = r.json() to happen ONLY if entity returns a 200. We want that even if entity returns a 400, and then the details of that 400 are added to the response object. 

Did I explain that well enough?